.github/workflows/failure-notification.yml

```diff
@@ -0,0 +1,88 @@
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+# DAMAGE.
+
+# Github Actions workflow to send out failure notification emails.
+# This is a reusable workflow that is called by various CI workflows.
+
+name: send-failure-notification
+on:
+  workflow_call:
+    inputs:
+      test_suite_status:
+        required: true
+        type: string
+      # We ask for the results only string separately so we can do a
+      # substring match to check for failure.
+      test_suite_results_only:
+        required: true
+        type: string
+    secrets:
+      DYNAMORIO_NOTIFICATION_EMAIL_USERNAME:
+        required: true
+      DYNAMORIO_NOTIFICATION_EMAIL_PASSWORD:
+        required: true
+
+jobs:
+  send-failure-notification-job:
+    runs-on: ubuntu-latest
+    # Reusable workflows are invoked as a separate job, which requires
+    # a new runner. But since this job is run only on master merge
+    # failures, which are expected to be rare, it is fine.
+    #
+    # We keep this condition in this reusable workflow to avoid its
+    # duplication across all callers. Callers may want to specify an
+    # always() condition to ensure this reusable workflow is not
+    # prematurely skipped. Note that the following condition still
+    # ensures that a new runner or job is not launched when not needed.
+    if: ${{ contains(inputs.test_suite_results_only, 'failure') && github.ref == 'refs/heads/master' }}
+    steps:
+    - name: Send failure mail to dynamorio-devs
+      uses: dawidd6/action-send-mail@v2
+      with:
+        server_address: smtp.gmail.com
+        server_port: 465
+        username: ${{secrets.DYNAMORIO_NOTIFICATION_EMAIL_USERNAME}}
+        password: ${{secrets.DYNAMORIO_NOTIFICATION_EMAIL_PASSWORD}}
+        subject: |
+          [${{github.repository}}] ${{github.workflow}} FAILED
+          on ${{github.event_name}} at ${{github.ref}}
+        body: |
+          Github Actions CI workflow run FAILED!
+          Workflow: ${{github.workflow}}
+          All test suite status: ${{ inputs.test_suite_status }}
+          Repository: ${{github.repository}}
+          Branch ref: ${{github.ref}}
+          SHA: ${{github.sha}}
+          Triggering actor: ${{github.actor}}
+          Triggering event: ${{github.event_name}}
+          Run Id: ${{github.run_id}}
+          See more details on github.com/DynamoRIO/dynamorio/actions/runs/${{github.run_id}}
```

Comment:
We should add text here requesting that the author take action and 1) ensure a bug is filed or existing bug has a xref added; 2) if a flake, re-run to get green; if not a flake, revert.  Right?

Reply:
Please see https://github.com/DynamoRIO/dynamorio/pull/7279#discussion_r1957245735. Or are you suggesting that we combine this PR (which is pure refactoring) with PR #7279 (which adds something new)?

Reply:
I would prefer to keep them separate.

Reply:
Oh ok, yes keep separate.